File: DFS/Lint_900.java
// https://www.jiuzhang.com/solution/closest-binary-search-tree-value/#tag-other

// M1. simple ver
public class Solution {
    /**
     * @param root: the given BST
     * @param target: the given target
     * @return: the value in the BST that is closest to the target
     */
    public int closestValue(TreeNode root, double target) {
        // write your code here
        int ans = root.val;
        
        while (root != null) {
            
            if (Math.abs(root.val - target) < Math.abs(ans - target)) {
                ans = root.val;
            }
            
            root = target <= root.val ? root.left : root.right;
        }
        
        return ans;
    }
}

File: 5 DFS_py/Lint_900.py
# M1, Binary Search Tree directly Traverse

class Solution:
    """
    @param root: the given BST
    @param target: the given target
    @return: the value in the BST that is closest to the target
    """
    def closestValue(self, root, target):
        if root is None:
            return None
        
        ans = float('inf')    
        while root:
            
            if abs(target - ans) > abs(target - root.val):
                ans = root.val
                
            root = root.left if target <= root.val else root.right
        
        return ans
